src/os/bluestore/ZonedAllocator.cc

```diff
@@ -164,6 +164,15 @@ bool ZonedAllocator::zoned_get_zones_to_clean(std::deque<uint64_t> *zones_to_cle
   return true;
 }
 
+void ZonedAllocator::zoned_mark_zone_clean(uint64_t zone_num) {
+  std::lock_guard l(lock);
+  auto removed = cleaning_in_progress_zones.erase(zone_num);
+  ceph_assert(removed);
+  zone_states[zone_num].num_dead_bytes = 0;
+  zone_states[zone_num].write_pointer = 0;
+  num_free += zone_size;
```

Comment:
Just in case - are you sure that we're are freeing the whole zone at this point?  Are we always going to clear totally filled zone? In other words - is it correct that num_free is unconditionally decremented by the full zone_size at this point. ZonedAllocator::allocate decrements num_free by want_size not zone_size and I'm having troubles in tracing num_free increment/decreament are in-sync.

Reply:
Hmm, good point. In theory, we should always clean a full zone, but given that we require the whole object to be in a single zone, we may have zones that are not completely full, with some small space at the end, and we may need to clean them. In that case we should increment num_free by the write_pointer of the zone. Fixed.